File: 009 Three Largest/Three_Largest.js
// O(n) Time | O(1) space
function findThreeLargestNumbers(array) {
	const threeLargest = [null, null, null];
	for (const num of array) {
		updateLargest(threeLargest, num);
	}
	return threeLargest;
}

function updateLargest(threeLargest, num) {
	if (threeLargest[2] === null || num > threeLargest[2]) {
		shiftAndUpdate(threeLargest, num, 2);
	} else if (threeLargest[1] === null || num > threeLargest[1]) {
		shiftAndUpdate(threeLargest, num, 1);
	} else if (threeLargest[0] === null || num > threeLargest[0]) {	
		shiftAndUpdate(threeLargest, num, 0);
	}
}

function shiftAndUpdate(array, num, index) {
	for (let i = 0; i <= index; i++) {
		if (i === index) {
			array[i] = num;
		} else {
			array[i] = array[i + 1];
		}
	}
}

File: 009 Three Largest/Three_Largest.py
def findThreeLargestNumbers(array):
  ## I just want to point out o how easy to do this in python it is 
  # array.sort()
  # return array[-3:]
  ## Now I gotta do it the "real" way
    largest = [None,None,None]
  for x in array:
    updateLargest(largest,x)
  return largest
  
  
def updateLargest(arr,x):
  if arr[2] is None or x > arr[2]:
    shiftAndUpdate(arr,x,2)
  elif arr[1] is None or x > arr[1]:
    shiftAndUpdate(arr,x,1)
  elif arr[0] is None or x > arr[0]:
    shiftAndUpdate(arr,x,0)
    
def shiftAndUpdate(arr, num, idx):
  for i in range(idx+1):
    if i == idx:
      arr[i] = num
    else:
      arr[i] = arr[i + 1]